File: model/src/main/java/no/nav/pensjon/vtp/testmodell/repo/impl/TestscenarioTemplateLoader.java
package no.nav.pensjon.vtp.testmodell.repo.impl;

import static java.nio.charset.StandardCharsets.UTF_8;
import static java.util.Arrays.stream;
import static java.util.function.Function.identity;
import static java.util.stream.Collectors.toMap;

import java.io.IOException;
import java.io.InputStream;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Stream;

import com.fasterxml.jackson.core.type.TypeReference;

import org.springframework.core.io.Resource;
import org.springframework.core.io.support.PathMatchingResourcePatternResolver;

import no.nav.pensjon.vtp.testmodell.repo.TestscenarioTemplate;
import no.nav.pensjon.vtp.testmodell.util.JsonMapper;
import no.nav.pensjon.vtp.testmodell.util.VariabelContainer;

public class TestscenarioTemplateLoader {
    public static final String LOCATION_PATTERN = "classpath:/scenarios/**/vars.json";

    private final JsonMapper jsonMapper = new JsonMapper();

    public Map<String, TestscenarioTemplate> load() {
        return locateVarsResources()
                .map(this::loadFileTestscenarioTemplate)
                .collect(toMap(TestscenarioTemplate::getTemplateKey, identity()));
    }

    private Stream<Resource> locateVarsResources() {
        try {
            return stream(new PathMatchingResourcePatternResolver().getResources(LOCATION_PATTERN));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private TestscenarioTemplate loadFileTestscenarioTemplate(Resource resource) {
        try {
            final String dir = resource.getURL().toString().substring(0, resource.getURL().toString().lastIndexOf("/"));
            final String templateName = URLDecoder.decode(dir.substring(dir.lastIndexOf("/") + 1), UTF_8);
            final VariabelContainer variables = loadTemplateVars(resource);
            return new FileTestscenarioTemplate(dir + "/", templateName, variables);
        } catch (final IOException e) {
            throw new IllegalStateException("Kunne ikke laste template fra " + resource, e);
        }
    }

    private VariabelContainer loadTemplateVars(Resource resource) throws IOException {
        try (InputStream is = resource.getInputStream()) {
            return new VariabelContainer(jsonMapper.lagObjectMapper().readValue(is, new TypeReference<HashMap<String, String>>() {}));
        }
    }
}

File: model/src/main/java/no/nav/pensjon/vtp/testmodell/repo/impl/TestscenarioTemplateLoader.kt
package no.nav.pensjon.vtp.testmodell.repo.impl

import com.fasterxml.jackson.core.type.TypeReference
import no.nav.pensjon.vtp.testmodell.repo.TestscenarioTemplate
import org.springframework.core.io.support.PathMatchingResourcePatternResolver
import java.net.URLDecoder
import no.nav.pensjon.vtp.testmodell.util.VariabelContainer
import no.nav.pensjon.vtp.testmodell.util.JsonMapper
import org.springframework.core.io.Resource
import java.nio.charset.StandardCharsets
import java.util.*

const val LOCATION_PATTERN = "classpath:/scenarios/**/vars.json"

class TestscenarioTemplateLoader {
    private val jsonMapper = JsonMapper()

    fun load(): Map<String, TestscenarioTemplate> {
        return locateVarsResources()
            .map { loadFileTestscenarioTemplate(it) }
            .associateBy({ it.templateKey }, { it })
    }

    private fun locateVarsResources(): List<Resource> {
        return PathMatchingResourcePatternResolver().getResources(LOCATION_PATTERN).toList()
    }

    private fun loadFileTestscenarioTemplate(resource: Resource): TestscenarioTemplate {
        val dir = resource.url.toString().substring(0, resource.url.toString().lastIndexOf("/"))
        val templateName = URLDecoder.decode(dir.substring(dir.lastIndexOf("/") + 1), StandardCharsets.UTF_8)
        val variables = loadTemplateVars(resource)
        return FileTestscenarioTemplate("$dir/", templateName, variables)
    }

    private fun loadTemplateVars(resource: Resource): VariabelContainer {
        resource.inputStream.use {
            return VariabelContainer(
                jsonMapper.lagObjectMapper()
                    .readValue<HashMap<String, String>>(it, object : TypeReference<HashMap<String, String>>() {})
            )
        }
    }
}
